src/wp-includes/media.php

```diff
@@ -1958,6 +1971,104 @@ function wp_img_tag_add_loading_attr( $image, $context ) {
+ */
+function wp_img_tag_add_loading_optimization_attrs( $image, $context ) {
+	$width              = preg_match( '/ width="([0-9]+)"/', $image, $match_width ) ? (int) $match_width[1] : '';
+	$height             = preg_match( '/ height="([0-9]+)"/', $image, $match_height ) ? (int) $match_height[1] : '';
+	$optimization_attrs = wp_get_loading_optimization_attributes(
+		'img',
+		array(
+			'width'  => $width,
+			'height' => $height,
+		),
+		$context
+	);
+
+	/**
+	 * Filters the loading optimization attributes to be added to an image.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string|bool $optimization_attrs   Attribute and it's values obtained from `wp_get_loading_optimization_attributes` function.
+	 * @param string      $image   The HTML `img` tag to be filtered.
+	 * @param string      $context Additional context about how the function was called or where the img tag is.
+	 */
+	$optimization_attrs = apply_filters( 'wp_img_tag_add_loading_optimization_attrs', $optimization_attrs, $image, $context );
+
+	/**
+	 * Filters the `loading` attribute value to add to an image. Default `lazy`.
+	 * This filter is added in for backward compatibility.
+	 *
+	 * Returning `false` or an empty string will not add the attribute.
+	 * Returning `true` will add the default value.
+	 * `true` and `false` usage supported for backward compatibility.
+	 *
+	 * @param string|bool $optimization_attrs['loading'] Current value for `loading` attribute for the image.
+	 * @param string      $image   The HTML `img` tag to be filtered.
+	 * @param string      $context Additional context about how the function was called or where the img tag is.
+	 */
+	$optimization_attrs['loading'] = apply_filters(
+		'wp_img_tag_add_loading_attr',
+		isset( $optimization_attrs['loading'] ) ? $optimization_attrs['loading'] : false,
+		$image,
+		$context
+	);
+
+	/**
+	 * Validate the values after filtering.
+	 */
+	if ( isset( $optimization_attrs['loading'] ) && $optimization_attrs['loading'] ) {
+		if ( ! in_array( $optimization_attrs['loading'], array( 'lazy', 'eager' ), true ) ) {
+			$optimization_attrs['loading'] = 'lazy';
+		}
+	}
+	if ( isset( $optimization_attrs['fetchpriority'] ) ) {
+		if ( ! in_array( $optimization_attrs['loading'], array( 'high', 'low', 'auto' ), true ) ) {
+			// For invalid values unset the attribute.
+			unset( $optimization_attrs['fetchpriority'] );
+		}
+	}
+
+	// Lazy-loading and fetchpriority='high' can't coexist at same time.
+	if (
+		isset( $optimization_attrs['loading'] ) && isset( $optimization_attrs['fetchpriority'] ) &&
+		'lazy' === $optimization_attrs['loading'] && 'high' === $optimization_attrs['fetchpriority']
+	) {
+		_doing_it_wrong(
+			__METHOD__,
+			__( 'An image cannot be lazy-loaded and assigned fetchpriority="high" at the same time.' ),
+			'6.3.0'
+		);
+		return $image;
+	}
+
+	// Retained for backward compatibility.
+	$add_img_loading_att = wp_lazy_loading_enabled( 'img', $context );
```

Comment:
@felixarntz @joemcgill  Even though our document suggested `The new function should also be called in each of the following functions, regardless of whether lazy-loading is enabled`. I have kept a enabled check withing the function itself (as some unit test were dependent on it). Let me know if i should keep it or removed it along with those unittest. Again visually nothing would break but my effect sites that are using it.

Reply:
We definitely have to keep the `wp_lazy_loading_enabled()` functionality working, so great point. However, having this here is still not all of it, because that function was previously also called in other places, like in `wp_get_attachment_image()` or `get_avatar()`. So I think to handle this holistically, we need to handle it everywhere where the new function `wp_get_loading_optimization_attributes()` is called.

So for example in `wp_get_attachment_image()` and `get_avatar()`, after calling the function, we could do something like:
```php
if ( isset( $loading_attrs['loading'] ) && ! wp_lazy_loading_enabled( 'img', $context ) ) {
	unset $loading_attrs['loading'];
}
```

I would recommend that you use this approach even here, right after the call to `wp_get_loading_optimization_attributes()`. This keeps the code consistent between where the function is called, and it's safer for backward compatibility as well: Keep in mind for example, that today, the `wp_img_tag_add_loading_attr` filter does not even fire when lazy-loading for images is generally disabled, but the way your code is right now it would still fire. So moving the check further up and not run the additional lazy-loading logic if it is disabled is more efficient and reliable.

Reply:
Agreed and modified the code.